File: MAP/lab5/src/models/expression/ArithExp.java
package models.expression;
import exceptions.*;
import utils.IDictionary;
import utils.IHeap;

public class ArithExp implements IExpression {

    private IExpression st, dr;
    private char op;

    public ArithExp(char _op, IExpression _st, IExpression _dr){
        op = _op;
        st = _st;
        dr = _dr;
    }


    @Override
    public int eval(IDictionary<String, Integer> dict, IHeap<Integer, Integer> heap) {


           int  vst = st.eval(dict, heap);
           int vdr = dr.eval(dict, heap);

          switch (op){
              case '+':
                  return vst + vdr;
              case '-':
                  return vst - vdr;
              case '*':
                  return vst * vdr;
              case '/':
                  if(vdr == 0)
                      throw new DivisionByZero("You have devided by 0");
                   return vst / vdr;
          }

          throw new InvalidOperator("Invalid Operator, please use only [+, -, *, /]");
    }

    @Override
    public String toString(){
        return "" + st + op + dr;
    }
}

File: MAP/lab7/lab7/Models/ArithExp.cs
using System;
using lab7.Utils;

namespace lab7.Models
{
    public class ArithExp : IExpression
    {
        private IExpression st;
        private IExpression dr;
        private char op;

        public ArithExp(IExpression st, IExpression dr, char op)
        {
            this.st = st;
            this.dr = dr;
            this.op = op;
        }


        public int eval(IDictionary<string, int> dict)
        {
            int vst = st.eval(dict);
            int vdr = dr.eval(dict);

            if (op == '+')
                return vst + vdr;
            if (op == '-')
                return vst - vdr;
            if (op == '*')
                return vst * vdr;
            else
            {
                if (vdr == 0)
                    throw new DivideByZeroException();
                else
                    return vst / vdr;
            }
        }
        
        public override string ToString()
        {
            return "" + st + op + dr;
        }




    }
}
